flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java

```diff
@@ -0,0 +1,370 @@
+		// before we can close the consumer and admin clients.
+		long closeTimeoutMs = KafkaSourceOptions.getOption(
+				properties,
+				KafkaSourceOptions.CLOSE_TIMEOUT_MS,
+				Long::parseLong);
+		close(closeTimeoutMs).ifPresent(t -> LOG.warn("Encountered error when closing KafkaSourceEnumerator", t));
+	}
+
+	@VisibleForTesting
+	Optional<Throwable> close(long timeoutMs) {
+		return closeWithTimeout(
+				"KafkaSourceEnumerator",
+				(ThrowingRunnable<Exception>) () -> {
+					consumer.close();
+					adminClient.close();
+				},
+				timeoutMs);
+	}
+
+	// ----------------- private methods -------------------
+
+	private PartitionSplitChange discoverAndInitializePartitionSplit() {
+		// Make a copy of the partitions to owners
+		KafkaSubscriber.PartitionChange partitionChange =
+				subscriber.getPartitionChanges(consumer, Collections.unmodifiableSet(discoveredPartitions));
+
+		Set<TopicPartition> newPartitions = Collections.unmodifiableSet(partitionChange.getNewPartitions());
+		OffsetsInitializer.PartitionOffsetsRetriever offsetsRetriever = getOffsetsRetriever();
+
+		Map<TopicPartition, Long> startingOffsets =
+				startingOffsetInitializer.getPartitionOffsets(newPartitions, offsetsRetriever);
+		Map<TopicPartition, Long> stoppingOffsets =
+				stoppingOffsetInitializer.getPartitionOffsets(newPartitions, offsetsRetriever);
+
+		Set<KafkaPartitionSplit> partitionSplits = new HashSet<>(newPartitions.size());
+		for (TopicPartition tp : newPartitions) {
+			Long startingOffset = startingOffsets.get(tp);
+			long stoppingOffset = stoppingOffsets.getOrDefault(tp, KafkaPartitionSplit.NO_STOPPING_OFFSET);
+			partitionSplits.add(new KafkaPartitionSplit(tp, startingOffset, stoppingOffset));
+		}
+		discoveredPartitions.addAll(newPartitions);
+		return new PartitionSplitChange(partitionSplits, partitionChange.getRemovedPartitions());
+	}
+
+	// This method should only be invoked in the coordinator executor thread.
+	private void handlePartitionSplitChanges(PartitionSplitChange partitionSplitChange, Throwable t) {
+		if (t != null) {
+			throw new FlinkRuntimeException("Failed to handle partition splits change due to ", t);
+		}
+		if (partitionDiscoveryIntervalMs < 0) {
+			LOG.debug("");
+			noMoreNewPartitionSplits = true;
+		}
+		// TODO: Handle removed partitions.
+		addPartitionSplitChangeToPendingAssignments(partitionSplitChange.newPartitionSplits);
+		assignPendingPartitionSplits();
+	}
+
+	// This method should only be invoked in the coordinator executor thread.
+	private void addPartitionSplitChangeToPendingAssignments(Collection<KafkaPartitionSplit> newPartitionSplits) {
+		int numReaders = context.currentParallelism();
+		for (KafkaPartitionSplit split : newPartitionSplits) {
+			// TODO: Implement a more sophisticated algorithm to reduce partition movement when parallelism changes.
+			int ownerReader = split.getTopicPartition().hashCode() % numReaders;
```

Comment:
would the hashCode() result in balanced assignment? should we reuse the `KafkaTopicPartitionAssigner` class? Also, if there is no unit test covering uniform distribution, we probably should add one.